### PR TITLE
Pass Authorization header

### PR DESCRIPTION
**Pull request by jr0d**
* Adds flag to enable passing the OIDC token via Authorization header
* Introduces the gorilla session API for preserving the token across
multiple redirects

### CODE REVIEW

File: internal/server.go [hunk 1]
```diff
@@ -50,22 +59,33 @@ func (s *Server) buildRoutes() {
...
+	})
+
 	// Modify request
 	r.Method = r.Header.Get("X-Forwarded-Method")
 	r.Host = r.Header.Get("X-Forwarded-Host")
 	r.URL, _ = url.Parse(GetUriPath(r))
 
+	if config.PassAuthorization {
+		logger.Debug("pass_authorization enabled, creating token session")
+		if err := s.createTokenSession(w, r); err != nil {
```
Comment (jkoelker) [inline internal/server.go]: Is creating the session at the root handler necessary? Couldn't the `saveToken` create it if not found, and the `getTokenFromSession` handle appropriately if it wasn't there?

Comment (jr0d) [reply internal/server.go]: It's not, now that passAuthorization is a configuration option. I originally tried to implement it as a query parameter, but traefik was filtering it out. So, good catch :)

File: internal/server.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package tfa
 
 import (
 	"fmt"
+	"github.com/gorilla/sessions"
```
Comment (mhrabovcin) [inline internal/server.go]: Usual golang import formatting is in the following order:

```
core

third-party

app packages
```

File: internal/server.go [hunk 3]
```diff
@@ -12,6 +12,12 @@ import (
 	"golang.org/x/oauth2"
 )
 
+const (
+	allAuthenticated       = "system:authenticated"
```
Comment (mhrabovcin) [inline internal/server.go]: Can you please document what is the meaning of this variable?

File: internal/config.go [hunk 4]
```diff
@@ -25,31 +25,34 @@ type Config struct {
...
+	CSRFCookieName          string               `long:"csrf-cookie-name" env:"CSRF_COOKIE_NAME" default:"_forward_auth_csrf" description:"CSRF Cookie Name"`
+	DefaultAction           string               `long:"default-action" env:"DEFAULT_ACTION" default:"auth" choice:"auth" choice:"allow" description:"Default action"`
+	Domains                 CommaSeparatedList   `long:"domain" env:"DOMAIN" description:"Only allow given email domains, can be set multiple times"`
+	LifetimeString          int                  `long:"lifetime" env:"LIFETIME" default:"43200" description:"Lifetime in seconds"`
+	Path                    string               `long:"url-path" env:"URL_PATH" default:"/_oauth" description:"Callback URL Path"`
+	SecretString            string               `long:"secret" env:"SECRET" description:"Secret used for signing (required)" json:"-"`
+	Whitelist               CommaSeparatedList   `long:"whitelist" env:"WHITELIST" description:"Only allow given email addresses, can be set multiple times"`
+	SessionKey              string               `long:"session-key" env:"SESSION_KEY" description:"Secret key used for signing and encrypting session cookies"`
```
Comment (mhrabovcin) [inline internal/config.go]: Can this be now removed?

File: go.mod [hunk 5]
```diff
@@ -12,7 +12,9 @@ require (
 	github.com/coreos/go-oidc v2.0.0+incompatible
 	github.com/go-acme/lego v2.5.0+incompatible // indirect
 	github.com/go-kit/kit v0.8.0 // indirect
+	github.com/golang/protobuf v1.3.2 // indirect
 	github.com/gorilla/context v1.1.1 // indirect
+	github.com/gorilla/sessions v1.2.0
```
Comment (mhrabovcin) [inline go.mod]: Is this necessary?

File: internal/server.go (hunk 3 above)
Comment (joejulian) [reply internal/server.go]: This is one of the default roles: https://kubernetes.io/docs/reference/access-authn-authz/rbac/#default-roles-and-role-bindings

**Pull request merged**